internal/store/elasticsearch/discovery_search_repository.go

```diff
@@ -277,3 +293,128 @@ func (repo *DiscoveryRepository) toSuggestions(response searchResponse) (results
 
 	return
 }
+
+func (repo *DiscoveryRepository) Group(ctx context.Context, cfg asset.GroupConfig) ([]asset.GroupResult, error) {
+	if len(cfg.GroupBy) == 0 || cfg.GroupBy[0] == "" {
+		err := asset.DiscoveryError{Err: fmt.Errorf("group by field cannot be empty")}
+		return nil, err
+	}
+	query, err := repo.buildGroupQuery(cfg)
+	if err != nil {
+		err = asset.DiscoveryError{Err: fmt.Errorf("error building query %w", err)}
+		return nil, err
+	}
+
+	res, err := repo.cli.client.Search(
+		repo.cli.client.Search.WithFilterPath("aggregations"),
+		repo.cli.client.Search.WithBody(query),
+		repo.cli.client.Search.WithIgnoreUnavailable(true),
+		repo.cli.client.Search.WithContext(ctx),
+	)
+
+	if err != nil {
+		err = asset.DiscoveryError{Err: fmt.Errorf("error executing group query %w", err)}
+		return nil, err
+	}
+
+	var response groupResponse
+
+	err = json.NewDecoder(res.Body).Decode(&response)
+	if err != nil {
+		fmt.Println(err.Error())
+		err = asset.DiscoveryError{Err: fmt.Errorf("error decoding group response %w", err)}
+		return nil, err
+	}
+	groupResultMap := make(map[string][]asset.Asset)
+	repo.ParseGroupResults(&response.Aggregations.TermAggregations, groupResultMap)
+	results := repo.toGroupResults(groupResultMap)
+	return results, nil
+}
+
+func (repo *DiscoveryRepository) ParseGroupResults(termAggregations *TermAggregations, groupResultMap map[string][]asset.Asset) {
+
+	for _, bucket := range termAggregations.Buckets {
+		if bucket.Group == nil {
+			key := bucket.Key
+			if assets, ok := groupResultMap[key]; ok {
+				groupResultMap[key] = append(assets, bucket.Hits.Hits.Hits[0].Source)
+			} else {
+				groupResultMap[key] = []asset.Asset{bucket.Hits.Hits.Hits[0].Source}
+			}
+		} else {
+			repo.ParseGroupResults(bucket.Group, groupResultMap)
+		}
+	}
+}
```

Comment:
Can you simplify this nested method ?

Reply:
Can't simplify it more. Added comments instead.

Reply:
This method has been changed.